feat: Allow custom serializers through `setSerializers`

alistairjcbrown opened the pull request:
Provides new `setSerializers` function to set the serializer objects to
be used before diffing. React component serializer implementation using
`react-test-renderer` has been extracted out.

Fixes #18
Fixes #30
Fixes #31

alistairjcbrown commented on __tests__/addSerializers.test.js:
```diff
@@ -0,0 +1,113 @@
+      }
+    )
+  ).toMatchSnapshot();
+});
```
Tests are a copy of those used for the react component testing in `snapshotDiff.test.js`

alistairjcbrown commented on src/react-serializer.js:
```diff
@@ -0,0 +1,43 @@
+      bAnnotation: prettyFormat(valueB, prettyFormatOptions),
+    };
+  },
+};
```
Code from `src/index.js` moved out to create a react component serializer object - this is currently the only default serializer

alistairjcbrown commented:
@thymikee I haven't added any doc updates as part of this PR, but give me a shout if the direction look ok and I'll get them added to the branch

thymikee commented on __tests__/addSerializers.test.js:
```diff
@@ -0,0 +1,113 @@
+        <span />
+        <span />
+        <span>{this.props.test}</span>
+        <span>{this.props.test}</span>
```
Both shallow and full render will look the same for this component, so this test doesn't really shows if this serializer works. I think instead of adding an example with massive `span`s amount, it would be better to create a `NestedComponent` nested inside these divs or spans, and see if they render shallowly. Or make it as another test case, to see if multiple serializers can live with each other

thymikee replied:
Could we also add a small test with React's shallow renderer? https://reactjs.org/docs/shallow-renderer.html

alistairjcbrown commented:
@thymikee Thanks - updated the branch 👍 
 - Added docs to readme
 - Tests updated to use nested component and now diff for fully rendered component, shallow rendered using react test renderer, shallow rendered using enzyme, POJO. Hopefully that shows off multiple serializers living together.
 - Looking at the issues I spotted #31, which notes wanting to be able to control how react components are rendered (amongst other things). As such I've updated this branch to instead `setSerializers, rather than `addSerializers` - the react serializer is now *not* included by default when calling `setSerializers`. This serializer is exposed via `snapshotDiff.reactSerializer` and can be added to the list of serializers passed into `snapshotDiff.setSerializers` by users.

thymikee commented:
Sorry for all the delay, I'll get to it next week. I'm pretty swamped with work right now 😅.

alistairjcbrown commented:
@thymikee No worries at all, give me a shout if there's anything else that needs updated 👍

alistairjcbrown commented:
@thymikee No pressure on this one - just making sure it hasn't fallen off the todo list

thymikee commented on README.md:
```diff
@@ -93,6 +93,41 @@ exports[`snapshot difference between 2 React components state 1`] = `
+const myCustomSerializer = require('./my-custom-serializer');
+
+snapshotDiff.setSerializers([
+  snapshotDiff.reactSerializer // use default react serializer - add this if you want to serialise react components!
```
let's export `defaultSerializers` object so that we can spread it here

thymikee commented on README.md:
```diff
@@ -93,6 +93,41 @@ exports[`snapshot difference between 2 React components state 1`] = `
 
+## Custom serializers
+
+By default, `snapshot-diff` uses a built in react serializer based on `react-test-renderer`. The serializers used can be
```
```suggestion
By default, `snapshot-diff` uses a built in React serializer based on `react-test-renderer`. The [serializers](https://jestjs.io/docs/en/configuration#snapshotserializers-array-string) used can be
```

thymikee commented on README.md:
```diff
@@ -93,6 +93,41 @@ exports[`snapshot difference between 2 React components state 1`] = `
+
+### Serializing react components with a different serializer
+
+You can replace the default react serializer by omitting it from the serializer list. The following uses enzymes to-json
```
```suggestion
You can replace the default React serializer by omitting it from the serializer list. The following uses enzymes to-json
```

thymikee commented on src/index.js:
```diff
@@ -29,12 +26,28 @@ const defaultOptions = {
 const SNAPSHOT_TITLE = 'Snapshot Diff:\n';
 
+const identity = value => value;
+let serializers = [reactSerializer];
```
Please export this property, so that users are able to import all the default serializers when calling `setSerializers`

thymikee commented on src/index.js:
```diff
@@ -29,12 +26,28 @@ const defaultOptions = {
-  if (isReactComponent(valueA) && isReactComponent(valueB)) {
-    difference = diffReactComponents(valueA, valueB, mergedOptions);
+  const matchingSerializer = serializers.find(
+    ({ test }) => test(valueA) && test(valueB)
```
Since we're now applying serializers one-by-one until we find a proper one, it would be nice to document that the order matters.

thymikee commented on src/index.js:
```diff
@@ -29,12 +26,28 @@ const defaultOptions = {
+    difference = diffStrings(
+      print(valueA, identity),
+      print(valueB, identity),
+      combinedOptions
```
Let's inline this, no need for an extra variable

thymikee commented:
@alistairjcbrown do you need any help to get this over the finishing line?

alistairjcbrown commented:
@thymikee Apologies, I hadn't caught up on my notifications. The branch has been updated - hopefully I've covered everything in the comments, but let me know if there's any more changes to do 👍

thymikee commented:
Thank you very much for this contribution. It's now released under v0.6.0 :)

SimenB commented:
Missing update to https://github.com/jest-community/snapshot-diff/blob/master/index.d.ts 😱
